services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,262 @@
+        // here Client Model of CD(Consumption Device) needs to be used to bind its Client Session with User Session.
+        AuthenticationSessionModel authSession = rootAuthSession.createAuthenticationSession(client);
+
+        authSession.setProtocol(OIDCLoginProtocol.LOGIN_PROTOCOL);
+        authSession.setAction(AuthenticatedClientSessionModel.Action.AUTHENTICATE.name());
+        authSession.setClientNote(OIDCLoginProtocol.ISSUER, Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+        authSession.setClientNote(OIDCLoginProtocol.SCOPE_PARAM, getScope());
+        logger.info(" specified scopes in backchannel authentication endpoint = " + getScope());
+
+        // authentication
+        AuthenticationFlowModel flow = AuthenticationFlowResolver.resolveCIBAFlow(authSession);
+        String flowId = flow.getId();
+        AuthenticationProcessor processor = new AuthenticationProcessor();
+        processor.setAuthenticationSession(authSession)
+                .setFlowId(flowId)
+                .setConnection(clientConnection)
+                .setEventBuilder(event)
+                .setRealm(realm)
+                .setSession(session)
+                .setUriInfo(session.getContext().getUri())
+                .setRequest(httpRequest);
+
+        processor.authenticateOnly();
+        processor.evaluateRequiredActionTriggers();
+        UserModel user = authSession.getAuthenticatedUser();
+        if (user.getRequiredActionsStream().count() > 0) {
+            event.error(Errors.RESOLVE_REQUIRED_ACTIONS);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "Account is not fully set up", Response.Status.BAD_REQUEST);
+        }
+
+        AuthenticationManager.setClientScopesInSession(authSession);
+
+        processor.attachSession();
+        UserSessionModel userSession = processor.getUserSession();
+        if (userSession == null) {
+            logger.info(" userSession is null.");
+        } else {
+            logger.info(" user session id = " + userSession.getId() + ", username = " + userSession.getUser().getUsername());
+        }
+        updateUserSessionFromClientAuth(userSession);
+
+        logger.info("  Created User Session's id                            = " + userSession.getId());
+        logger.info("  Submitted in advance User Session ID Will Be Created = " + getUserSessionIdWillBeCreated());
+
+        // authorization (consent)
+        UserConsentModel grantedConsent = session.users().getConsentByClient(realm, user.getId(), client.getId());
+        if (grantedConsent == null) {
+            grantedConsent = new UserConsentModel(client);
+            session.users().addConsent(realm, user.getId(), grantedConsent);
+            logger.info("  Consent updated : ");
+            grantedConsent.getGrantedClientScopes().stream().forEach(i->logger.info(i.getName()));
```

Comment:
can be simplified to `grantedConsent.getGrantedClientScopes().forEach(i->logger.info(i.getName()));`

Reply:
@valb3r Yes, I will fix codes as you suggested.